Rethinking_2/Chp_07.ipynb

```diff
@@ -0,0 +1,3517 @@
+{
```

Comment:
<p>Not exactly the same not sure if nicer :-( but maybe it helps you to think of something better</p><p> </p><p>s0 = az.compare(dataset_dict)['dse']</p><p>del dataset_dict[s0.index[0]]</p><p>s1 = az.compare(dataset_dict)['dse']</p><p>df_0 = s0.to_frame(name=s0.index[0])</p><p>df_1 = s1.to_frame(name=s1.index[0])</p><p> </p><p>pd.merge(df_0 , df_1, left_index=True, right_index=True)</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/resources/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='77'/>

Reply:
Thanks @aloctavodia . What is `dataset_dict` in your code?

Reply:
ohh, sorry about that `dataset_dict` is 

ms = {'m_6_6' : m_6_6_trace,
      'm_6_7' : m_6_7_trace,
      'm_6_8' : m_6_8_trace}

Reply:
Thanks! I've implemented this with some explanatory notes, and wrapped it in a function in case people want to apply it to more than three models.

Still feels a bit hack-y, but much nicer than mine! Would be nice if `arviz` could do this... Might submit a feature request.

Reply:
For reference, the function:

```python
def pairwise_compare(dataset_dict, metric='dse', **kwargs):
    """
    Calculate pairwise comparison of models in dataset_dict.
    
    Parameters
    ----------
    dataset_dict : dict
        A dict containing two ore more {'name': pymc3.backends.base.MultiTrace}
        items.
    metric : str
        The name of the matric to be calculated. Can be any valid column output
        by `arviz.compare`. Note that this may change depending on the **kwargs
        that are specified.
    kwargs
        Arguments passed to `arviz.compare`
    """
    data_dict = dataset_dict.copy()
    dicts = []
    
    while len(data_dict) > 1:
        c = az.compare(data_dict, **kwargs)[metric]
        dicts.append(c.to_frame(name=c.index[0]))
        del data_dict[c.index[0]]
    
    return pd.concat(dicts, axis=1).dropna(how='any')
```

Reply:
<p>I am not sure the current version of the code does the same as the first. Originally all calls to compare were done with pairs of models, now, isn't the first call done with 3 models and the 2nd with 2?</p>

Reply:
As I understood it, in @aloctavodia's original the first call compares all three models, and `compare` returns stats comparing two models to the 'best' model (`s0`). The 'best' model is then dropped (the `del` call), and then the remaining two models are compared. As long as the comparison metric is symmetric, this is valid. In my function I've applied the same logic, running sequential `compare` calls then deleting the 'best' model in each iteration, which has the same effect (the outputs of the original and my function are identical). The benefit of the function is that it works for *n* models.

For reference, @aloctavodia's code with some annotation:

```python
dataset_dict = {"m_6_6": m_6_6_trace, "m_6_7": m_6_7_trace, "m_6_8": m_6_8_trace}

# compare all models
s0 = az.compare(dataset_dict, ic="waic", scale="deviance")["dse"]
# the output compares each model to the 'best' model - i.e. two models are compared to one.
# to complete a pair-wise comparison we need to compare the remaining two models.

# to do this, remove the 'best' model from the input data
del dataset_dict[s0.index[0]]

# re-run compare with the remaining two models
s1 = az.compare(dataset_dict, ic="waic", scale="deviance")["dse"]

# s0 compares two models to one model, and s1 compares the remaining two models to each other
# now we just nee to wrangle them together!

# convert them both to dataframes, setting the name to the 'best' model in each `compare` output.
# (i.e. the name is the model that others are compared to)
df_0 = s0.to_frame(name=s0.index[0])
df_1 = s1.to_frame(name=s1.index[0])

# merge these dataframes to create a pairwise comparison
pd.merge(df_0, df_1, left_index=True, right_index=True)
```

All of this is in the [Chp_07 notebook](https://github.com/oscarbranson/resources/blob/Chp_07/Rethinking_2/Chp_07.ipynb) of the branch this PR is from (Code 7.31).

There is, however, a slight error in the function. The `dropna` call at the end should be removed.